### DISCUSSION TITLE
GIF decoding issue

### DISCUSSION THREAD

**Discussion by maleblond** [libvips/libvips#2698] [Help wanted]
Hi!

Thank you so much for the recent improvements to the GIF encoding!

While testing those changes, we noticed libvips / libnsgif is having trouble decoding certain GIFs while the same GIFs can be successfully decoded with gifsicle. Here is an example:

<img src="https://user-images.githubusercontent.com/11162235/156180232-dc5ac6ef-c5b5-45e6-a0a2-1c4de5075ba4.gif" width=300>

I did some debugging and it seems like there is a `0` byte right after the end of frame 34, and before the start of frame 35. The GIF decoder instead expect a byte that represents the start of a new block (I think either 0x21 or 0x2c, as explained [there](http://giflib.sourceforge.net/whatsinagif/bits_and_bytes.html)). libnsgif just returns an error in this case (https://github.com/libvips/libvips/blob/master/libvips/foreign/libnsgif/libnsgif.c#L866), but gifsicle has some tolerance (https://github.com/kohler/gifsicle/blob/15c36c2f2c113c21650fa81a6381841ca1536c59/src/gifread.c#L880-L888).

I tested that skipping this `0` byte after the 34th frame allows libnsgif to properly decode the GIF. Something like adding this block right there https://github.com/libvips/libvips/blob/master/libvips/foreign/libnsgif/libnsgif.c#L1168:

```
	// Some malformed GIFs seem to have junk data between blocks that we tolerate
	// (up to 20 bytes of junk data between the last processed block and the current one)
	int bad_block_id_count = 0;
	while (pos[0] != GIF_EXTENSION_INTRODUCER && pos[0] != GIF_IMAGE_SEPARATOR) {
		pos++;
		if(pos >= end || bad_block_id_count++ == 20) {
			return GIF_DATA_ERROR;
		}
	}
```

~This actually lead to a segfault (not sure why yet)~(EDIT: the segfault I've been experimenting seems to be due to something unrelated). The idea would be to skip unexpected byte as gifsicle does.

I'm not sure if this case is actually compliant to the GIF spec or not, but in all cases I think doing a best effort to find the next block start doesn't really hurt. Just dropping the frame may also be a good tradeoff.

We'd be curious to know what you think about this!

**Comment by jcupitt** [comment]
Hi @maleblond, thanks for the report.

This is some way beyond my GIF knowledge heh. Can I ping you @tlsa ? Have you come across issues like this?

fwiw, git master libvips shows:

```
$ vipsheader garden.gif 
vipsheader: gifload: GIF frame data error
```

I had a look at gif master libnsgif, but it looks like the API has changed and it's not easy for me to test.

**Reply by tlsa** [reply]
Thanks for the report. I'll look into it.

@maleblond do you have any more example GIFs like this?

@jcupitt I'll update libvips to the new libnsgif API and open a MR.

**Reply by maleblond** [reply]
Hi @tlsa !

Here's another example:

<img src="https://user-images.githubusercontent.com/11162235/156211830-bb0ba556-6140-4a3f-ae1a-b9cfe9ac17cd.gif" width="300">

**Reply by tlsa** [reply]
Thanks!

**Comment by tlsa** [comment]
Should be fixed by https://github.com/libvips/libvips/pull/2702.
